### PR TITLE
bpo-14074: argparse doesn't allow metavar to be a tuple

### PR DESCRIPTION
**Pull request by cykerway**
Currently argparse allows nargs>1 for positional arguments but doesn't
allow metavar to be a tuple:

    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('foo', nargs=2, metavar=('bar', 'baz'))
    args = parser.parse_args()

The above code will raise exception both with and without '--help'.

This patch fixed some functions about metavar processing so the above
code will pass.

https://bugs.python.org/issue14074

* Issue: gh-58282

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by embray** [comment]
Thanks for the PR.  I can confirm that this fixes this (very old) bug...

**Comment by rhettinger** [comment]
Consider restricting this to the case where *nargs* is an integer.  It doesn't make much sense with ``*`` or ``REMAINDER``.

File: Lib/argparse.py [hunk 1]
```diff
@@ -709,6 +711,9 @@ def _get_action_name(argument):
     elif argument.option_strings:
         return  '/'.join(argument.option_strings)
     elif argument.metavar not in (None, SUPPRESS):
+        if isinstance(argument.nargs, int):
+            if isinstance(argument.metavar, tuple):
```
Comment (ZackerySpytz) [inline Lib/argparse.py]: Please combine these `if`s.

File: Misc/NEWS.d/next/Library/2018-12-04-07-36-27.bpo-14074.fMLKCu.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fix argparse metavar processing to allow positional arguments to have a
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Library/2018-12-04-07-36-27.bpo-14074.fMLKCu.rst]:
```suggestion
Fix :mod:`argparse` metavar processing to allow positional arguments to have a
```

File: Lib/argparse.py [hunk 3]
```diff
@@ -724,6 +726,9 @@ def _get_action_name(argument):
     elif argument.option_strings:
         return  '/'.join(argument.option_strings)
     elif argument.metavar not in (None, SUPPRESS):
+        if isinstance(argument.nargs, int) and \
+           isinstance(argument.metavar, tuple):
+            return ' '.join(argument.metavar)
```
Comment (jacobtylerwalls) [inline Lib/argparse.py]: CPython follows PEP8, so I believe the style is to prefer open parentheses over `\`.

Comment (cykerway) [reply Lib/argparse.py]: There it is: https://github.com/python/cpython/pull/10847/commits/c0204fdc3f5c2f58b738f88ccf638cb388a117af

Nothing else changed.

**Comment by rhettinger** [comment]
For:

     parser.add_argument('foo', nargs=2, metavar=('bar', 'baz'))

Wouldn't it be better to specify two separate arguments?

     parser.add_argument('foo')
     parser.add_argument('baz')

Having two different meta variables implies that the arguments have different meanings.  Collapsing them into a single argument seems like an anti-pattern to me.  Both warrant their own error messages, possible type conversions, and possible default values.

File: Lib/argparse.py [hunk 4]
```diff
@@ -704,6 +706,9 @@ def _get_action_name(argument):
     elif argument.option_strings:
         return '/'.join(argument.option_strings)
     elif argument.metavar not in (None, SUPPRESS):
+        if isinstance(
+            argument.nargs, int) and isinstance(argument.metavar, tuple):
+            return ' '.join(argument.metavar)
```
Comment (serhiy-storchaka) [inline Lib/argparse.py:711]: This change is not covered by tests.

File: Lib/test/test_argparse.py [hunk 5]
```diff
@@ -4665,6 +4665,28 @@ class TestHelpTupleMetavar(HelpTestCase):
     version = ''
 
 
+class TestHelpTupleMetavarPositional(HelpTestCase):
+    """Test specifying metavar on a Positional as a tuple"""
+
+    parser_signature = Sig(prog='PROG')
+    argument_signatures = [
+        Sig('foo', help='foo help', nargs=2, metavar=('bar', 'baz'))
```
Comment (serhiy-storchaka) [inline Lib/test/test_argparse.py:4673]:
```suggestion
        Sig('w', help='w help', nargs='+', metavar=('W1', 'W2')),
        Sig('x', help='x help', nargs='*', metavar=('X1', 'X2')),
        Sig('y', help='y help', nargs=3, metavar=('Y1', 'Y2', 'Y3')),
        Sig('z', help='z help', nargs='?', metavar=('Z1', )),
```

File: Lib/test/test_argparse.py [hunk 6]
```diff
@@ -4665,6 +4665,28 @@ class TestHelpTupleMetavar(HelpTestCase):
...
+    ]
+    argument_group_signatures = []
+    usage = '''\
+        usage: PROG [-h] bar baz
```
Comment (serhiy-storchaka) [inline Lib/test/test_argparse.py:4677]:
```suggestion
        usage: PROG [-h] W1 [W2 ...] [X1 [X2 ...]] Y1 Y2 Y3 [Z1]
```

File: Lib/test/test_argparse.py [hunk 7]
```diff
@@ -4665,6 +4665,28 @@ class TestHelpTupleMetavar(HelpTestCase):
...
+        '''
+    help = usage + '''\
+
+        positional arguments:
+          bar baz     foo help
```
Comment (serhiy-storchaka) [inline Lib/test/test_argparse.py:4682]:
```suggestion
          W1 W2       w help
          X1 X2       x help
          Y1 Y2 Y3    y help
          Z1          z help
```

File: Lib/argparse.py [hunk 8]
```diff
@@ -528,6 +528,8 @@ def _format_action(self, action):
     def _format_action_invocation(self, action):
         if not action.option_strings:
             default = self._get_default_metavar_for_positional(action)
+            if isinstance(action.nargs, int):
+                return ' '.join(self._metavar_formatter(action, default)(1))
             metavar, = self._metavar_formatter(action, default)(1)
             return metavar
```
Comment (serhiy-storchaka) [inline Lib/argparse.py:534]:
```suggestion
            return ' '.join(self._metavar_formatter(action, default)(1))
```

**Review by serhiy-storchaka — COMMENTED** [review]
I wondered why `TestHelpTupleMetavarPositional` tests only argument with numerical value of `nartgs`, while `TestHelpTupleMetavar` tests with all `nargs` values. So I tried to change the test. It failed, but a simple change in `_format_action_invocation()` can fix this. I am nor sure that this is a best representation for `nargs='*'` and `nargs='+'`, but this at least fixes failure.

The change in `_get_action_name()` is not tested. We need a test for it, and it should be generalized to support other `nargs` values.

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @cykerway and sorry for a lack of response.
#124782 generalizes this PR to support other nargs values.

**Pull request closed without merging**